src/main/scala/rise/core/types/Solution.scala

```diff
@@ -0,0 +1,134 @@
+  def subs(aa: AddressSpaceIdentifier, ab: AddressSpace): Solution =
+    Solution(Map(), Map(), Map(aa -> ab), Map(), Map(), Map())
+  def subs(na: NatToDataIdentifier, nb: NatToData): Solution =
+    Solution(Map(), Map(), Map(), Map(na -> nb), Map(), Map())
+  def subs(na: NatToNatIdentifier, nb: NatToNat): Solution =
+    Solution(Map(), Map(), Map(), Map(), Map(na -> nb), Map())
+  def subs(na: NatCollectionIdentifier, nb: NatCollection): Solution =
+    Solution(Map(), Map(), Map(), Map(), Map(), Map(na -> nb))
+}
+
+case class Solution(ts: Map[Type, Type],
+                    ns: Map[NatIdentifier, Nat],
+                    as: Map[AddressSpaceIdentifier, AddressSpace],
+                    n2ds: Map[NatToDataIdentifier, NatToData],
+                    n2ns: Map[NatToNatIdentifier, NatToNat],
+                    natColls: Map[NatCollectionIdentifier, NatCollection]
+                   ) {
+  import traversal.{Continue, Result, Stop}
+
+  case class Visitor(sol: Solution) extends traversal.Visitor {
+    override def visitNat(ae: Nat): Result[Nat] = Stop(sol(ae))
+    override def visitType[T <: Type](t: T): Result[T] =
+      Stop(sol(t).asInstanceOf[T])
+    override def visitAddressSpace(a: AddressSpace): Result[AddressSpace] =
+      Stop(sol(a))
+    override def visitN2D(n2d: NatToData): Result[NatToData] = Stop(sol(n2d))
+
+    override def visitN2N(n2n: NatToNat): Result[NatToNat] = Stop(sol(n2n))
+  }
+
+  def apply(e: Expr): Expr = {
+    traversal.DepthFirstLocalResult(e, Visitor(this))
+  }
+
+  def apply(t: Type): Type = {
+    traversal.types.DepthFirstLocalResult(t, new Visitor(this) {
+      override def visitType[T <: Type](t: T): Result[T] =
+        sol.ts.get(t) match {
+          case Some(x) => Stop(x.asInstanceOf[T])
+          case None    => Continue(t, this)
+        }
+    })
+  }
+
+  def apply(n: Nat): Nat =
+    (substitute.natsInNat(ns, _)).andThen(substitute.n2nsInNat(n2ns, _))(n)
+
+
+  def apply(a: AddressSpace): AddressSpace =
+    substitute.addressSpacesInAddressSpace(as, a)
+
+  def apply(a: NatCollection): NatCollection =
+    substitute.natCollectionInNatCollection(natColls, a)
+
+  def apply(n2d: NatToData): NatToData = {
+    substitute.n2dsInN2d(n2ds, n2d) match {
+      case id: NatToDataIdentifier => id
+      case lambda@NatToDataLambda(x, body) =>
+        val xSub = apply(x) match {
+          case n: NatIdentifier => n
+          case other => throw NonIdentifierInBinderException(lambda, other)
+        }
+        NatToDataLambda(xSub, apply(body).asInstanceOf[DataType])
+    }
+  }
+
+  def apply(n2n: NatToNat): NatToNat = {
+    substitute.n2nsInN2n(n2ns, n2n) match {
+      case id: NatToNatIdentifier => id
+      case lambda@NatToNatLambda(x, body) =>
+        val xSub = apply(x) match {
+          case n: NatIdentifier => n
+          case other => throw NonIdentifierInBinderException(lambda, other)
+        }
+        NatToNatLambda(xSub, apply(body))
+    }
+  }
```

Comment:
For the most part, no. NatToNat and NatToData only appear when DepArrays/DepPairs are involved, and so those code paths shouldn't be taken in non-dependent programs.

There is one exception to this though -  I had to add a pass in the solution of a dep constraint, that basically does the following: for a dependent function n:K -> T, traverse T, and for all type identifiers in T, replace them with the appropriate function dependent on n. So say, if you meet the dependent function n:Nat -> n.t, with t being a type variable, this will replace t with f(n), where f is a NatToData function. This is because any type variable could be really a dependent function applied, and this dependence needs to be tracked explicitly there and then - or else it will be lost in further substitutions.

Reply:
ok, we can see how the CI times will change on shine for example when things are integrated.